badger/badger.go

```diff
@@ -0,0 +1,228 @@
+package badger
+
+import (
+	"bytes"
+	"encoding/binary"
+
+	"github.com/dgraph-io/badger"
+	"github.com/pkg/errors"
+	"github.com/smallstep/nosql"
+)
+
+// DB is a wrapper over *badger.DB,
+type DB struct {
+	db *badger.DB
+}
+
+// Open opens or creates a BoltDB database in the given path.
+func (db *DB) Open(_opts interface{}) (err error) {
+	opts, ok := _opts.(badger.Options)
+	if !ok {
+		return errors.Errorf("DB.Open expects badger.Options, but got %T", _opts)
+	}
+	db.db, err = badger.Open(opts)
+	return errors.WithStack(err)
+}
+
+// Close closes the DB database.
+func (db *DB) Close() error {
+	return errors.WithStack(db.db.Close())
+}
+
+// CreateTable is a noop because badger does not support tables, buckets,
+// collections, etc.
+func (db *DB) CreateTable(bucket []byte) error {
+	return nil
+}
+
+// DeleteTable deletes a root or embedded bucket. Returns an error if the
+// bucket cannot be found or if the key represents a non-bucket value.
+func (db *DB) DeleteTable(bucket []byte) error {
+	return errors.New("error: DeleteTable not supported by DB")
+}
+
+func badgerGet(txn *badger.Txn, key []byte) (ret []byte, err error) {
+	item, err := txn.Get(key)
+	if err != nil {
+		if err == badger.ErrKeyNotFound {
+			return nil, errors.WithStack(nosql.ErrNotFound)
+		}
+		return nil, errors.WithStack(err)
+	}
+
+	val, err := item.Value()
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	// Make sure to return a copy as ret is only valid during the
+	// transaction.
+	ret = cloneBytes(val)
+	return
+}
+
+// Get returns the value stored in the given bucked and key.
+func (db *DB) Get(bucket, key []byte) (ret []byte, err error) {
+	bk, err := toBadgerKey(bucket, key)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	err = db.db.View(func(txn *badger.Txn) error {
+		ret, err = badgerGet(txn, bk)
+		return err
+	})
+	return
+}
+
+func badgerSet(txn *badger.Txn, key, value []byte) error {
+	return errors.WithStack(txn.Set(key, value))
+}
+
+// Set stores the given value on bucket and key.
+func (db *DB) Set(bucket, key, value []byte) error {
+	bk, err := toBadgerKey(bucket, key)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	return db.db.Update(func(txn *badger.Txn) error {
+		return badgerSet(txn, bk, value)
```

Comment:
The number of lines of the implementation of badgetSet is also one, you can just use something like:
```go
return errors.Wrapf(txn.Set(bk, value), "failed to set %s/%s", bucket, key)
```